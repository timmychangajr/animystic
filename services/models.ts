export interface Image {
  id: string;
  description: string;
  updated_at: string;
  urls: {
    raw: string;
    full: string;
    regular: string;
    small: string;
    thumb: string;
  };
  user: {
    username: string;
    name: string;
  };
  tags: string[];
}

export interface ImageResponse {
  success: boolean;
  data: {
    total: number;
    total_pages: number;
    current_page: number;
    per_page: number;
    results: Image[];
  };
}

export interface Animal {
  imageUrl?: string,
  characteristics?: {
    name_of_young?: string;
    litter_size?: string;
    average_litter_size?: string,
    color?: string,
    diet?: string,
    favorite_food?: string,
    habitat?: string,
    lifespan?: string,
    lifestyle?: string,
    main_prey?: string,
    prey?: string,
    predators?: string,
    skin_type?: string,
    slogan?: string,
    top_speed?: string,
    type?: string,
    weight?: string
  },
  locations?: string[],
  name: string,
  taxonomy?: {
    class?: string,
    family?: string,
    genus?: string,
    kingdom?: string,
    order?: string,
    phylum?: string,
    scientific_name?: string
  }
}