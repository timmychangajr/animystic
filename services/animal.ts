import getImage from "./image";

export interface Animal {
  imageUrl?: string,
  characteristics?: {
    name_of_young?: string;
    litter_size?: string;
    average_litter_size?: string,
    color?: string,
    diet?: string,
    favorite_food?: string,
    habitat?: string,
    lifespan?: string,
    lifestyle?: string,
    main_prey?: string,
    prey?: string,
    predators?: string,
    skin_type?: string,
    slogan?: string,
    top_speed?: string,
    type?: string,
    weight?: string
  },
  locations?: string[],
  name: string,
  taxonomy?: {
    class?: string,
    family?: string,
    genus?: string,
    kingdom?: string,
    order?: string,
    phylum?: string,
    scientific_name?: string
  }
}

export default async function getAnimal(query: string): Promise<Animal | undefined> {
  const apiKey = process.env.EXPO_PUBLIC_API_NINJAS_KEY ?? '';
  let animal;
  try {
    await fetch('https://api.api-ninjas.com/v1/animals?name=' + query, {
      method: 'GET',
      headers: {
        'X-Api-Key': apiKey,
        'Content-Type': 'application/json'
      },
    }).then(async data => {
      const res = await data.json()
      animal = res[0] as Animal;
      if (animal?.name) animal.imageUrl = await getImage(animal.name)
    })
  } catch (e) {
    console.error('error getting animal', e)
  }
  return animal;
}