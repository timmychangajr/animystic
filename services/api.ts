import { Animal, ImageResponse } from "./models";

export async function getAnimal(query: string): Promise<Animal | undefined> {
  const apiKey = process.env.EXPO_PUBLIC_API_NINJAS_KEY ?? '';
  let animal;
  try {
    await fetch('https://api.api-ninjas.com/v1/animals?name=' + query, {
      method: 'GET',
      headers: {
        'X-Api-Key': apiKey,
        'Content-Type': 'application/json'
      },
    }).then(async data => {
      const res = await data.json()
      animal = res[0] as Animal;
      if (animal?.name) animal.imageUrl = await getImage(animal.name)
    })
  } catch (e) {
    console.error('error getting animal', e)
  }
  return animal;
}

export async function getImage(query: string): Promise<string | undefined> {
  const apiKey = process.env.EXPO_PUBLIC_RAPID_API_KEY ?? '';
  let image;
  try {
    await fetch('https://unsplash-image-search-api.p.rapidapi.com/search?page=1&query=' + query, {
      method: 'GET',
      headers: {
        'x-rapidapi-key': apiKey,
        'x-rapidapi-host': 'unsplash-image-search-api.p.rapidapi.com',
        'Content-Type': 'application/json',
      },
    }).then(async data => {
      const res = await data.json() as ImageResponse
      image = res.data.results[0].urls.regular
    })
  } catch (e) {
    console.error('error getting image', e)
  }
  return image;
}