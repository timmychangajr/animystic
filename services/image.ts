
interface Image {
  id: string;
  description: string;
  updated_at: string;
  urls: {
    raw: string;
    full: string;
    regular: string;
    small: string;
    thumb: string;
  };
  user: {
    username: string;
    name: string;
  };
  tags: string[];
}

interface ImageResponse {
  success: boolean;
  data: {
    total: number;
    total_pages: number;
    current_page: number;
    per_page: number;
    results: Image[];
  };
}

export default async function getImage(query: string): Promise<string | undefined> {
  const apiKey = process.env.EXPO_PUBLIC_RAPID_API_KEY ?? '';
  let image;
  try {
    await fetch('https://unsplash-image-search-api.p.rapidapi.com/search?page=1&query=' + query, {
      method: 'GET',
      headers: {
        'x-rapidapi-key': apiKey,
        'x-rapidapi-host': 'unsplash-image-search-api.p.rapidapi.com',
        'Content-Type': 'application/json',
      },
    }).then(async data => {
      const res = await data.json() as ImageResponse
      image = res.data.results[0].urls.regular
    })
  } catch (e) {
    console.error('error getting books', e)
  }
  return image;
}