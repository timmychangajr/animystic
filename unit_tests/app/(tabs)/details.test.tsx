import DetailsScreen from '@/app/(tabs)/details';
import { Animal } from '@/services/animal';
import { cleanup, render } from '@testing-library/react-native';
import { useLocalSearchParams } from 'expo-router';

jest.mock('expo-router', () => {
  return {
    ...jest.requireActual('expo-router'),
    useLocalSearchParams: jest.fn(),
  };
});

describe('DetailsScreen', () => {
  afterEach(cleanup);
  const initSearchParams = async (params?: object) => {
    useLocalSearchParams.mockImplementation(() => ({
      fromNav: params ? JSON.stringify(params): 'undefined'
    }))
  }
  test('renders correctly', async () => {
    await initSearchParams()
    const tree = render(<DetailsScreen />);
    expect(tree).toMatchSnapshot();
  });
  test('should load valid details with an image', async () => {
    await initSearchParams({
      name: 'Sloth',
      imageUrl: 'test'
    } as Animal)
    const tree = render(<DetailsScreen />);
    expect(await tree.getByTestId('animal_name')).toBeOnTheScreen()
  })
  test('should load valid details without an image', async () => {
    await initSearchParams({
      name: 'Sloth',
      imageUrl: 'test'
    } as Animal)
    const tree = render(<DetailsScreen />);
    expect(await tree.getByTestId('animal_name')).toBeOnTheScreen()
  })
});