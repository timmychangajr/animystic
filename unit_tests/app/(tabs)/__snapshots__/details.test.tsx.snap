// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsScreen renders correctly 1`] = `
<View
  style={
    [
      false,
      {
        "backgroundColor": "#000000B5",
      },
      undefined,
    ]
  }
>
  <ActivityIndicator
    color="#CCCCCC"
    size="large"
  />
</View>
`;
